.github/workflows/release.yml

```
@@ -293,6 +296,105 @@ jobs:
+        shell: bash
+        run: |
+          set -ex
+          commit_id=$(git rev-parse HEAD)
+          echo "commit_sha=$commit_id" >> $GITHUB_OUTPUT
+          filename=reproduce-${{ matrix.platform.name }}.json
+
+          cat <<EOF > $filename
+          {
+            "platform_name": "${{ matrix.platform.name }}",
+            "build_container_image": "${{needs.image_digest.outputs.image_digest}}",
+            "tdnf_snapshottime": $SOURCE_DATE_EPOCH,
+            "commit_sha": "$commit_id"
+          }
+          EOF
+          cat $filename
+
+      - name: "Upload Reproducibility Metadata"
+        uses: actions/upload-artifact@v4
+        with:
+          name: reproduce-${{ matrix.platform.name }}-metadata
+          path: reproduce-${{ matrix.platform.name }}.json
+
+  reproduce_rpm:
+    name: Test Reproducibility
+    needs:
+      - build_release
+      - image_digest
+    strategy:
+      matrix:
+        platform:
+          - name: virtual
+            os: azure-linux
+          - name: snp
+            os: azure-linux
+
+    runs-on: [self-hosted, 1ES.Pool=gha-virtual-ccf-sub]
+    env:
+      OUTPUT_DIR: /tmp/reproduced
+    container:
+      image: ${{ needs.image_digest.outputs.image_digest }}
+      volumes:
+        - /tmp/reproduced:/tmp/reproduced
+      options: "--user root --publish-all --cap-add NET_ADMIN --cap-add NET_RAW --cap-add SYS_PTRACE"
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - name: Download Prebuilt Packages
+        uses: actions/download-artifact@v4
+        with:
+          path: PREBUILT_DIR
+          pattern: pkg-*
+          merge-multiple: true
+      - name: Download Reproducibility Metadata
+        uses: actions/download-artifact@v4
+        with:
+          path: repro
+          pattern: reproduce-${{ matrix.platform.name }}-metadata
+          merge-multiple: true
+
+      - name: Reproduce RPMs
+        id: reproduce
+        shell: bash
+        run: |
+          set -ex
+          chmod +x ./reproduce/reproduce_rpm.sh
+          ./reproduce/reproduce_rpm.sh repro/reproduce-${{ matrix.platform.name }}.json
+          reproduced_pkg=$(ls "$OUTPUT_DIR" | head -n1 || true)
+          if [ -z "$reproduced_pkg" ]; then
+            echo "ERROR: No reproduced package found in $OUTPUT_DIR"
+            exit 1
+          fi
+          echo "Reproduced package: $reproduced_pkg"
+          echo "reproduced_pkg=$reproduced_pkg" >> $GITHUB_OUTPUT
+
+      - name: Verify reproducibility
+        id: verify
+        shell: bash
+        run: |
+          set -ex
+          echo "Verifying reproducibility..."
+          reproduced_pkg=${{ steps.reproduce.outputs.reproduced_pkg }}
+          if ! diff "$OUTPUT_DIR/$reproduced_pkg" "PREBUILT_DIR/$reproduced_pkg"; then
+            echo "Failed to reproduce $reproduced_pkg!"
+            exit 1
+          else
+            echo "Reproducibility verified for $reproduced_pkg"
+          fi
```

```suggestion
        run: cmp $PREBUILT_DIR/${{ steps.reproduce.outputs.reproduced_pkg }} $OUTPUT_DIR/$reproduced_pkg
```

The names are fairly confusing, I wonder if we can make them clearer.